language_models: Add support for tool use to LM Studio provider

gyzerok opened the pull request:
Closes #30004

**Quick demo:**

https://github.com/user-attachments/assets/0ac93851-81d7-4128-a34b-1f3ae4bcff6d

**Additional notes:**

I've tried to stick to existing code in OpenAI provider as much as possible without changing much to keep the diff small.

This PR is done in collaboration with @yagil from LM Studio. We agreed upon the format in which LM Studio will return information about tool use support for the model in the upcoming version. As of current stable version nothing is going to change for the users, but once they update to a newer LM Studio tool use gets automatically enabled for them. I think this is much better UX then defaulting to true right now.


Release Notes:

- Added support for tool calls to LM Studio provider

gyzerok committed:
add support for tool use to LM Studio provider

gyzerok commented:
@yagil am happy to look into it, if you can share this LM Studio build with me

yagil commented:
> @yagil am happy to look into it, if you can share this LM Studio build with me

@gyzerok thank you! Build 3 available here: https://lmstudio.ai/beta-releases

The relevant changelog item is

```
- LM Studio REST API (/api/v0): return model capabilities in GET /models response
  - e.g. `"capabilities": ["tool_use"]`
```

gyzerok commented:
@yagil just tested it with couple different models and everything seems to work fine for me. Maybe you aren't using the latest commit from my branch?

yagil commented:
> @yagil just tested it with couple different models and everything seems to work fine for me. Maybe you aren't using the latest commit from my branch?

Ah that’s very possible. Glad it’s working!

gyzerok committed:
add tool choice support

gyzerok committed:
avoid selecting default_model and default_fast_model

yagil commented:
Latest commit from this PR + LM Studio 0.3.16b3 works!

<img width="578" alt="image" src="https://github.com/user-attachments/assets/4fb60ce4-4dd5-4a09-a33b-24f6e03e585c" />

Previously I had accidentally used `main` from the fork instead of the feature branch, btw.

gyzerok committed:
take into account new image tool result

gyzerok committed:
fix after rebase to the current main

benbrandt committed:
Refactor LmStudioEventMapper to avoid unnecessary clones

Replace borrowing operations with moves where possible to improve
performance. This eliminates redundant cloning of strings and other
data when processing response stream events.

benbrandt commented:
@gyzerok and @yagil thanks for all of the work here! It looks good to me, will merge once the CI passes

gyzerok commented:
@benbrandt thank you for looking into this! I have some other improvements, which I've decided to make in a separate PRs to ease the review process. Seems like I can go on with them now :)

sebastienbarre commented:
Howdy. I'm a bit confused by this. I'm using LM Studio with qwen2.5-coder-7b-instruct-mlx. I connected to it from Zed, but the Agent panel shows "Tools Unsupported", so I can't select "write". Nevertheless, if I ask "what tools do you have?", it will show a list, including "edit_file", and if I ask it to create a file according to some spec, it will do it. But shouldn't it display the list of tools?
